### ISSUE TITLE
Functorch memory_efficient_fusion gives wrong output batch size

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#86020]
### 🐛 Describe the bug

Run the following script:

```python
import torch

from functorch.compile import memory_efficient_fusion

import timm

x = torch.randn(5, 3, 224, 224).cuda()
y = torch.randn(6, 3, 224, 224).cuda()

net = timm.create_model('resnet50').cuda()
net = memory_efficient_fusion(net)
net.eval()

with torch.jit.fuser('fuser2'):
    xout = net(x)
    yout = net(y)

print(xout.shape, yout.shape)
```

Got:
```
torch.Size([5, 1000]) torch.Size([5, 1000])
```

Expecting:
```
torch.Size([5, 1000]) torch.Size([6, 1000])
```

Note that `yout` has the wrong batch size. Commenting out the memory_efficient_fusion (using eager mode) gives the expected result.

### Versions

Was working fine on 9/28

torch_commit: https://github.com/pytorch/pytorch/commit/614d6f19e3d30cac0d77059e738d1f25d75eb408
torchdynamo_commit: https://github.com/pytorch/torchdynamo/commit/6ead5cae0d1234aa64db06fe230ef56e12ec76fe
torchvision_commit: https://github.com/pytorch/vision/commit/a35be97a6e6725c83032315a8f5e5f6911c9ef41

First saw error on 9/29

torch_commit: https://github.com/pytorch/pytorch/commit/3cfc61b84659cea435411a546eca6a891584247f
torchdynamo_commit: https://github.com/pytorch/torchdynamo/commit/6ead5cae0d1234aa64db06fe230ef56e12ec76fe
torchvision_commit: https://github.com/pytorch/vision/commit/a35be97a6e6725c83032315a8f5e5f6911c9ef41

**Comment by Chillee** [comment]
Oh, we deleted the cache on AOTAutograd recently, so it doesn't recompile when the shape changes. The idea here is that AOTAutograd is intended to sit behind torchdynamo, which will handle all the caching, so the caching in AOTAutograd is basically unneeded code.

**Comment by Chillee** [comment]
@xwang233 is there any reason why you're using AOTAutograd directly instead of behind torchdynamo?

**Comment by xwang233** [comment]
Using torchdynamo does give the expected result.

```python
import torch
import torchdynamo

import timm

x = torch.randn(5, 3, 224, 224).cuda()
y = torch.randn(6, 3, 224, 224).cuda()

net = timm.create_model('resnet50').cuda()
net = torchdynamo.optimize('aot_nvfuser')(net)
net.eval()

xout = net(x)
yout = net(y)

print(xout.shape, yout.shape)
```

```
torch.Size([5, 1000]) torch.Size([6, 1000])
```

**Comment by xwang233** [comment]
@Chillee Is there any document update that recommend users to not use `memory_efficient_fusion` directly?

**Comment by Chillee** [comment]
Yeah, the issue here is that since AOTAutograd specializes on static shapes today, it needs to retrace/recompile when the input shapes change. We used to have a caching layer there, but we recently removed it due to causing extra complexity: https://github.com/pytorch/pytorch/pull/85783

You can still use `memory_efficient_fusion` directly today, you just need to call it again on the model when the shapes change.

> We currently still have memory_efficient_fusion testing on TIMM models. Should we replace that with torchdynamo's aot_autograd?

Yeah, that's the recommended path forward.

**Comment by xwang233** [comment]
Thank you for the update. I think proper documentations on those would be much nice.

cc @csarofeen @kevinstephano

**Comment by Chillee** [comment]
tbh, if it's problematic, we could probably bring the cache back.
